src/mlpack/methods/ann/layer/maxout_impl.hpp

```diff
@@ -0,0 +1,134 @@
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_MAXOUT_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_MAXOUT_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "maxout.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+Maxout<InputDataType, OutputDataType>::Maxout()
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+Maxout<InputDataType, OutputDataType>::Maxout(
+    const size_t inSize,
+    const size_t hiddenSize,
+    const size_t outSize) :
+    inSize(inSize),
+    hiddenSize(hiddenSize),
+    outSize(outSize)
+{
+  weights.set_size((inSize * hiddenSize * outSize) + (outSize * hiddenSize), 1);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void Maxout<InputDataType, OutputDataType>::Reset()
+{
+  weight = arma::mat(weights.memptr(), outSize * hiddenSize, inSize,
+      false, false);
+  bias = arma::mat(weights.memptr() + weight.n_elem,
+      outSize * hiddenSize, 1, false, false);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Maxout<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (output.n_elem == 0)
+    output = arma::zeros<arma::mat>(outSize, input.n_cols);
+
+  hiddenParameter = (weight * input) + bias;
+
+  for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++)
+    output.row(j) = arma::max(hiddenParameter.rows(i, i + hiddenSize - 1), 0);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Maxout<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& /* input */, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  arma::mat hiddenGradient = arma::zeros<arma::mat>(hiddenParameter.n_rows,
+      hiddenParameter.n_cols);
+  arma::mat hiddenParameterSubset;
+
+  for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++)
+  {
+    hiddenParameterSubset = hiddenParameter.rows(i, i + hiddenSize - 1);
+    
+    for (int k = 0; k < hiddenParameterSubset.n_cols; k++)
+    {
+      int ind = arma::as_scalar(arma::find(arma::max(
+          hiddenParameterSubset.col(k)) == hiddenParameterSubset.col(k)));
+      hiddenGradient(i + ind, k) = gy(j, k);
+    }
+  }
+
+  g = weight.t() * hiddenGradient;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Maxout<InputDataType, OutputDataType>::Gradient(
+    const arma::Mat<eT>&& input,
+    arma::Mat<eT>&& error,
+    arma::Mat<eT>&& gradient)
+{
+  arma::mat hiddenGradient = arma::zeros<arma::mat>(hiddenParameter.n_rows,
+      hiddenParameter.n_cols);
+  arma::mat hiddenParameterSubset;
+
+  for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++)
+  {
+    hiddenParameterSubset = hiddenParameter.rows(i, i + hiddenSize - 1);
+    
+    for (int k = 0; k < hiddenParameterSubset.n_cols; k++)
+    {
+      int ind = arma::as_scalar(arma::find(arma::max(
```

Comment:
Hi @zoq I bench marked the maxout code with 3 different implementations which are:
**maxoutIndices**: Storing index of maximum hidden units and then using it.
**maxoutValues**: Storing value of maximum hidden unit and then using it (basically storing output of maxout layer).
**hiddenValues**: current approach I have implemented.

I tested it by training end to end model for all 3 setups (since this changes affected all forward backward and gradient passes). 

<table>
<tr>
<th>method</th>
<th>iterations</th>
<th>time</th>
</tr>

<tr>
<td>maxoutIndices</td>
<td>70</td>
<td>55100 ms </td>
</tr>
<tr>
<td>maxoutIndices</td>
<td>7</td>
<td>4552.45 ms</td>
</tr>
<tr>
<td>maxoutValues</td>
<td>70</td>
<td>57945.5 ms</td>
</tr>
<tr>
<td>maxoutValues</td>
<td>7</td>
<td>4346.81 ms</td>
</tr>
<tr>
<td>hiddenParameters</td>
<td>70</td>
<td>52324.9 ms</td>
</tr>
<tr>
<td>hiddenParameters</td>
<td>7</td>
<td>4373.22 ms</td>
</tr>
</table>

Honestly I expected slight performance boost from maxoutIndices and maxoutValues method but from observations it seems that hiddenParameters outperforms both of them. I am confused by this result. So I leave this decision on you, too.